src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ManagedHResultExceptionMarshallerFactory.cs

```diff
@@ -0,0 +1,110 @@
+        {
+            if (info.MarshallingAttributeInfo is ManagedHResultExceptionMarshallingInfo)
+            {
+                if (_direction == MarshalDirection.UnmanagedToManaged)
+                {
+                    return new UnmanagedToManagedMarshaller();
+                }
+                return new ManagedToUnmanagedMarshaller();
+            }
+            else
+            {
+                return _inner.Create(info, context);
+            }
+        }
+
+        private sealed class ManagedToUnmanagedMarshaller : IMarshallingGenerator
+        {
+            public ManagedTypeInfo AsNativeType(TypePositionInfo info) => info.ManagedType;
+            public IEnumerable<StatementSyntax> Generate(TypePositionInfo info, StubCodeContext context)
+            {
+                Debug.Assert(info.MarshallingAttributeInfo is ManagedHResultExceptionMarshallingInfo);
+
+                if (context.CurrentStage != StubCodeContext.Stage.Unmarshal)
+                {
+                    yield break;
+                }
+
+                (_, string nativeIdentifier) = context.GetIdentifiers(info);
+
+                // Marshal.ThrowExceptionForHR(<native>);
+                yield return ExpressionStatement(
+                    InvocationExpression(
+                        MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression,
+                            ParseName(TypeNames.System_Runtime_InteropServices_Marshal),
+                            IdentifierName("ThrowExceptionForHR")),
+                        ArgumentList(
+                            SingletonSeparatedList(Argument(IdentifierName(nativeIdentifier))))));
+            }
+
+            public SignatureBehavior GetNativeSignatureBehavior(TypePositionInfo info) => SignatureBehavior.NativeType;
+            public ValueBoundaryBehavior GetValueBoundaryBehavior(TypePositionInfo info, StubCodeContext context) => ValueBoundaryBehavior.ManagedIdentifier;
+            public bool IsSupported(TargetFramework target, Version version) => true;
+            public bool SupportsByValueMarshalKind(ByValueContentsMarshalKind marshalKind, StubCodeContext context) => false;
+            public bool UsesNativeIdentifier(TypePositionInfo info, StubCodeContext context) => false;
+        }
+
+        private sealed class UnmanagedToManagedMarshaller : IMarshallingGenerator
+        {
+            public ManagedTypeInfo AsNativeType(TypePositionInfo info) => info.ManagedType;
+            public IEnumerable<StatementSyntax> Generate(TypePositionInfo info, StubCodeContext context)
+            {
+                Debug.Assert(info.MarshallingAttributeInfo is ManagedHResultExceptionMarshallingInfo);
+
+                if (context.CurrentStage != StubCodeContext.Stage.Unmarshal)
+                {
+                    yield break;
+                }
+
+                (_, string nativeIdentifier) = context.GetIdentifiers(info);
```

Comment:
So both `ManagedToUnmanaged` and `UnmanagedToManaged` are using `nativeIdentifier`?

Reply:
Since `UsesNativeIdentifier` returns false, they should actually both use the managed identifier (in practice the two are the same for the cases where this marshaller is used). I'll update these to use to the managed identifier.